File: python/open_guardrail/guards/response_caching_safe.py
"""Detect responses unsafe for caching."""

import re
import time
from typing import List

from open_guardrail.core import GuardResult

_PII = [
    re.compile(r"\b\d{3}[-.]?\d{2}[-.]?\d{4}\b"),
    re.compile(
        r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+"
        r"\.[A-Z]{2,}\b",
        re.I,
    ),
]

_TIMESTAMPS = [
    re.compile(r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}"),
    re.compile(r"\d{1,2}:\d{2}\s*(?:AM|PM)\b", re.I),
]

_USER_SPECIFIC = [
    re.compile(r"\byour\s+account\b", re.I),
    re.compile(r"\byour\s+order\b", re.I),
    re.compile(r"\byour\s+balance\b", re.I),
]

_SESSION = [
    re.compile(r"\bsession[_\s-]?id\b", re.I),
    re.compile(r"\blogged\s+in\s+as\b", re.I),
]


class _ResponseCachingSafe:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "response-caching-safe"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        reasons: List[str] = []
        for p in _PII:
            if p.search(text):
                reasons.append("pii")
                break
        for p in _TIMESTAMPS:
            if p.search(text):
                reasons.append("timestamp")
                break
        for p in _USER_SPECIFIC:
            if p.search(text):
                reasons.append("user-specific")
                break
        for p in _SESSION:
            if p.search(text):
                reasons.append("session-ref")
                break
        triggered = len(reasons) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="response-caching-safe",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=(
                f"Unsafe for caching: {', '.join(reasons)}"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"reasons": reasons} if triggered else None
            ),
        )


def response_caching_safe(
    *, action: str = "warn"
) -> _ResponseCachingSafe:
    return _ResponseCachingSafe(action=action)

File: packages/guards/src/response-caching-safe.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

interface ResponseCachingSafeOptions {
  action: 'block' | 'warn';
}

const PII_PATTERNS = [
  /\b\d{3}[-.]?\d{2}[-.]?\d{4}\b/,
  /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z]{2,}\b/i,
  /\b\d{4}[\s-]?\d{4}[\s-]?\d{4}[\s-]?\d{4}\b/,
];

const TIMESTAMP_PATTERNS = [
  /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}/,
  /\b(?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\s+\d{1,2},?\s+\d{4}\b/i,
  /\d{1,2}:\d{2}\s*(?:AM|PM)\b/i,
];

const USER_SPECIFIC = [
  /\byour\s+account\b/i,
  /\byour\s+order\b/i,
  /\byour\s+balance\b/i,
  /\byour\s+profile\b/i,
  /\byour\s+subscription\b/i,
];

const SESSION_REFS = [
  /\bsession[_\s-]?id\b/i,
  /\bsession[_\s]?token\b/i,
  /\blogged\s+in\s+as\b/i,
  /\bcurrent\s+user\b/i,
];

export function responseCachingSafe(options: ResponseCachingSafeOptions): Guard {
  return {
    name: 'response-caching-safe',
    version: '0.1.0',
    description: 'Detect responses unsafe for caching',
    category: 'security',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const reasons: string[] = [];

      for (const p of PII_PATTERNS) {
        if (p.test(text)) { reasons.push('pii'); break; }
      }
      for (const p of TIMESTAMP_PATTERNS) {
        if (p.test(text)) { reasons.push('timestamp'); break; }
      }
      for (const p of USER_SPECIFIC) {
        if (p.test(text)) { reasons.push('user-specific'); break; }
      }
      for (const p of SESSION_REFS) {
        if (p.test(text)) { reasons.push('session-ref'); break; }
      }

      const triggered = reasons.length > 0;

      return {
        guardName: 'response-caching-safe',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        message: triggered
          ? `Unsafe for caching: ${reasons.join(', ')}`
          : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { reasons } : undefined,
      };
    },
  };
}
